src/ass/to_ass.py

```diff
@@ -1,84 +1,242 @@
-        format=pyass.EventFormat.DIALOGUE
-        if shouldPrintTitle
-        else pyass.EventFormat.COMMENT,
-        style=TITLE_STYLE.name,
-        end=TITLE_EVENT_DURATION,
-        parts=[
-            pyass.EventPart(tags=TITLE_CARD_TAGS, text=r"\N".join(s)),
-        ],
+def get_enter_transition_tag(
+    kChar: KChar,
+    switchDuration: timedelta,
+    transitionDuration: timedelta,
+    resultTag: Sequence[pyass.Tag] = [pyass.AlphaTag(0x00)],
+) -> pyass.Tag:
+    return pyass.TransformTag(
+        start=kChar.fadeOffset, end=switchDuration + kChar.fadeOffset, to=resultTag
     )
 
 
-def to_events(
-    song: Song,
+def get_exit_transition_tag(
+    kChar: KChar,
+    switchDuration: timedelta,
+    transitionDuration: timedelta,
+    resultTag: Sequence[pyass.Tag] = [pyass.AlphaTag(0xFF)],
+) -> pyass.Tag:
+    return pyass.TransformTag(
+        start=kChar.line.duration
+        - transitionDuration
+        + switchDuration
+        + kChar.fadeOffset,
+        end=kChar.line.duration
+        - transitionDuration
+        + 2 * switchDuration
+        + kChar.fadeOffset,
+        to=resultTag,
+    )
+
+
+def get_char_actor_tag(
+    kChar: KChar,
     actorToStyle: Mapping[str, Sequence[pyass.Tag]],
-    shouldPrintTitle: bool = True,
-    switchDuration: timedelta = DEFAULT_SWITCH_DURATION,
-    transitionDuration: timedelta = DEFAULT_TRANSITION_DURATION,
-) -> Sequence[pyass.Event]:
-    return [
-        to_divider_event(song, song.title.romaji),
-        to_title_event(song, shouldPrintTitle),
-        to_divider_event(song, "Romaji"),
-        *[
-            to_default_event(
-                to_romaji_k_line(line, i + 1),
-                actorToStyle,
-                switchDuration,
-                transitionDuration,
-            )
-            for i, line in enumerate(song.lyrics)
-        ],
-        to_divider_event(song, "English"),
-        *[
-            to_default_event(
-                to_en_k_line(line, i + 1),
-                actorToStyle,
-                switchDuration,
-                transitionDuration,
-            )
-            for i, line in enumerate(song.lyrics)
-        ],
+    switchDuration: timedelta,
+) -> Sequence[pyass.Tag]:
+    if len(kChar.line.actorSwitches) == 0:
+        return []
+    return list(actorToStyle[kChar.line.startActor]) + [
+        pyass.TransformTag(
+            start=time + kChar.fadeOffset,
+            end=time + kChar.fadeOffset,
+            to=actorToStyle[actor],
+        )
+        for time, actor in kChar.line.actorSwitches
     ]
+
+
+def get_char_karaoke_tag(
+    kChar: KChar, resultTag: Sequence[pyass.Tag] = []
+) -> Sequence[pyass.Tag]:
+    if kChar.karaDuration == timedelta():
+        return []
+    # Karaoke timing
+    if len(resultTag) == 0:
+        return [pyass.KaraokeTag(kChar.karaDuration)]
+    else:
+        return [
+            pyass.TransformTag(start=kChar.karaStart, end=kChar.karaEnd, to=resultTag)
+        ]
```

Comment:
Do you see these methods as belonging here or in `effect/default_effect.py`?

I think things like `get_exit_transition_tag` seem quite specific to the default effect and is unlikely to see any re-use in other user-defined effects, but if you think otherwise, feel free to ignore this comment. 

If the functions are only relevant to `default_effect`, though, I think they can be moved to `default_effect.py` as well?

Reply:
Good point, I can't think of any instance where that would be the case. Updated.